module/data/item/spell.mjs

```diff
@@ -335,6 +335,46 @@ export default class SpellData extends ItemDataModel.mixin(ActivitiesTemplate, I
     return data;
   }
 
+  /* -------------------------------------------- */
+  /*  Socket Event Handlers                       */
+  /* -------------------------------------------- */
+
+  /** @inheritDoc */
+  _preCreate(data, options, user) {
+    if ( super._preCreate(data, options, user) === false ) return false;
+    if ( !this.parent.isEmbedded || ["atwill", "innate"].includes(this.preparation.mode) || this.sourceClass ) return;
+    const classes = new Set(Object.keys(this.parent.actor.spellcastingClasses));
+    if ( !classes.size ) return;
+
+    // Set the source class, and ensure the preparation mode is pact if adding a prepared spell to a pact class
```

Comment:
This can be generalized to any "alternative" spell slot type.

Reply:
Give my new PR a look to make sure that will work with alt types: https://github.com/foundryvtt/dnd5e/pull/4068